### PR TITLE
temp_conversion

### LINKED ISSUE #30: Convert the Temperature

### PR DESCRIPTION
**Pull request by Elshikh-Amro**
---
name: solution review 
about: A template PR for code review with a checklist
---

### Files

- [x] The file name describes the function's behavior
- [x] There is a module docstring in the function file
- [x] The test file's name matches the function file name -
  `/tests/test_file_name.py`
- [x] There is a module docstring in the tests file

### Unit Tests

- [x] The test class has a helpful name in PascalCase
- [x] The test class has a docstring
- Every unit test has
  - [x] A helpful name
  - [x] A clear docstring
  - [x] Only one assertion
  - [x] There is no logic in the unit test
- [x] All tests pass
- [x] There are tests for defensive assertions
- [x] There are tests for boundary cases

### Function Docstring

- [x] The function's behavior is described
- The function's arguments are described:
  - [x] Type
  - [x] Purpose
  - [x] Other assumptions (eg. if it's a number, what's the expected range?)
- The return value is described
  - [x] Type
  - [x] Other assumptions are documented
- The defensive assertions are documented using `Raises:`
  - [x] Each assumption about an argument is checked with an assertion
  - [x] Each assertion checks for _only one_ assumption about the argument
- [x] Include 3 or more (passing!) doctests

### The Function

- [x] The function's name describes it's behavior
- [x] The function's name matches the file name
- [x] The function has correct type annotations
- [x] The function is not called in the function file

### Do's

- [x] Variable names help to understand the strategy
- [x] Any comments are clear and describe the strategy
- [x] Lines of code are spaced to help show different stages of the strategy

### Don'ts

- [x] The function's strategy _is not_ described in the documentation
- [x] Comments explain the _strategy_, **not** the _implementation_
- [x] The function _does not_ have more comments than code
  - If it does, consider finding a new strategy or a simpler implementation

## Implementation

- [x] The code passes the formatting checks
- [x] The code passes all Ruff linting checks
- [x] The code has no (reasonable) Pylint errors
  - In code review, you can decide when fixing a Pylint error is helpful and
    when it's too restricting.
- [x] Variables are named with snake_case
- [x] Variable names are clear and helpful
- [x] The code follows the strategy as simply as possible
- [x] The implementation is as simple as possible given the strategy
- [x] There are no commented lines of code
- [x] There are no `print` statements anywhere
- [x] The code includes defensive assertions
- [x] Defensive assertions include as little logic as possible

### CODE REVIEW

File: solutions/temp_conversion.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+def convert_temperature(celsius):
```
Comment (SiSaR-Pal) [inline solutions/temp_conversion.py]: Update the function signature to include type annotations:

```suggestion
def convert_temperature(celsius: float) -> dict[str, float]:


```

File: solutions/temp_conversion.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        >>> convert_temperature(0)
+        {'Fahrenheit': 32.0, 'Kelvin': 273.15}
+        >>> convert_temperature(-273.15)
+        {'Fahrenheit': -459.67, 'Kelvin': 0.0}
+        >>> convert_temperature(100)
+        {'Fahrenheit': 212.0, 'Kelvin': 373.15}
+        >>> convert_temperature(-300)  # Below absolute zero
+        Traceback (most recent call last):
+        ValueError: Temperature cannot be below absolute zero (-273.15°C).
+    """
```
Comment (SiSaR-Pal) [inline solutions/temp_conversion.py]: Currently, the function assumes that celsius is always a number (float). 
Validate the type of the input and raise a TypeError if it isn't a number.
```suggestion
    """
        if not isinstance(celsius, (int, float)):
        raise TypeError("The input temperature must be a number.")
```

**Review by SiSaR-Pal — COMMENTED** [review]
The code and tests are well-written, maintainable, and align with the PR template. The function is robust, and the unit tests thoroughly cover typical cases, edge cases, and invalid inputs. Excellent work!

File: solutions/temp_conversion.py (hunk 1 above)
Comment (Elshikh-Amro) [reply solutions/temp_conversion.py]: Ok that is really helpful

File: solutions/temp_conversion.py (hunk 2 above)
Comment (Elshikh-Amro) [reply solutions/temp_conversion.py]: Yessss i missed that

**Comment by Elshikh-Amro** [comment]
it says no print statement but my function return more than one argument am i allowed to bypass this law or should i change my function

**Comment by Elshikh-Amro** [comment]
@SiSaR-Pal @CaesarGhazi @Nada-saad635 @omniaNS @AseelOmer @heba3

**Comment by Nada-saad635** [comment]
no worries Amro cause your code return and there is no print statement for it , the rest of the code is good and well written ! nice work you can merge it

**Comment by Nada-saad635** [comment]
no worries Amro cause your code return and there is no print statement for it , the rest of the code is good and well written ! nice work you can merge it

**Review by Nada-saad635 — COMMENTED** [review]
no worries Amro cause your code return and there is no print statement for it , the rest of the code is good and well written ! nice work you can merge it

File: solutions/temp_conversion.py [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+        raise ValueError("Temperature cannot be below absolute zero (-273.15°C).")
+
+    fahrenheit = (celsius * 9 / 5) + 32
+    kelvin = celsius + 273.15
+
+    return {"Fahrenheit": round(fahrenheit, 2), "Kelvin": round(kelvin, 2)}
+
+
+def main():
+    print("Welcome to the Temperature Converter!")
```
Comment (Nada-saad635) [inline solutions/temp_conversion.py]: here its not important you can delete it

File: solutions/temp_conversion.py [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    try:
+        user_input = input("Enter a temperature in Celsius: ")
+        celsius = float(user_input)
+        result = convert_temperature(celsius)
+        print(f"The temperature in Fahrenheit is {result['Fahrenheit']}°F.")
+        print(f"The temperature in Kelvin is {result['Kelvin']}K.")
```
Comment (Nada-saad635) [inline solutions/temp_conversion.py]: here  convert it to return both print

File: solutions/temp_conversion.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    except ValueError as e:
+        print(f"Error: {e}")
```
Comment (Nada-saad635) [inline solutions/temp_conversion.py]: here too convert it to return

Comment (Elshikh-Amro) [reply solutions/temp_conversion.py]: done

File: solutions/temp_conversion.py (hunk 4 above)
Comment (Elshikh-Amro) [reply solutions/temp_conversion.py]: done

File: solutions/temp_conversion.py (hunk 3 above)
Comment (Elshikh-Amro) [reply solutions/temp_conversion.py]: done

**Review by omniaNS — CHANGES_REQUESTED** [review]
Good job Amro! A few things to improve for checklist compliance:
- add module docstrings to both the function file and the test file
- function name should match the file name (unify the names; convert_temperature vs temp_conversion)
- main function calls convert_temperature, consider removing main(), as the function shouldn't be called in the function file according to the checklist
- add comments to explain the function's strategy

**Comment by Elshikh-Amro** [comment]
@omniaNS can you check again

**Review by omniaNS — COMMENTED** [review]
Thanks for addressing the feedback @Elshikh-Amro Just a few more tweaks and you are ready to go 🚀:
- I noticed a small typo in the updated file names, make sure the word "temperature" is spelled correctly so the function and file names match exactly.
👉 Pro Tip: consider using the [Code Spell Checker](https://marketplace.visualstudio.com/items?itemName=streetsidesoftware.code-spell-checker) VSCode extension to catch typos like this, it’s very useful!
- remove the function docstring from test file
- each file should start with a module docstring (before any code) to explain its overall purpose, examples:

  for the function file:
  
  ```
  """
  This module provides functionality to convert temperatures from Celsius
  to Fahrenheit and Kelvin.
  functions:
      - convert_temperature(celsius: float) -> dict[str, float]
  """
  ```
   for the test file:
   ```
   """
  Unit tests for the convert_temperature function.
  """
  ```
  you can also add author name and date of creation (if you want)
  Author: Amro
  Created: January 2, 2025

**Comment by Elshikh-Amro** [comment]
thank you @omniaNS

File: solutions/tests/test_convert_temperature.py [hunk 6]
```diff
@@ -1,29 +1,9 @@
...
-        {'Fahrenheit': 32.0, 'Kelvin': 273.15}
-        >>> convert_temperature(-273.15)
-        {'Fahrenheit': -459.67, 'Kelvin': 0.0}
-        >>> convert_temperature(100)
-        {'Fahrenheit': 212.0, 'Kelvin': 373.15}
-        >>> convert_temperature(-300)  # Below absolute zero
-        Traceback (most recent call last):
-        ValueError: Temperature cannot be below absolute zero (-273.15°C).
-    """
+   Unit tests for the convert_temperature function.
```
Comment (omniaNS) [inline solutions/tests/test_convert_temperature.py]: the module docstring goes at the very top before your import statement

Comment (Elshikh-Amro) [reply solutions/tests/test_convert_temperature.py]: ok

File: solutions/convert_temperature.py [hunk 7]
```diff
@@ -10,7 +10,8 @@ def convert_temperature(celsius: float) -> dict[str, float]:
 
     Raises:
         ValueError: If the input is not a valid number or is below absolute zero.
-
+    functions:
```
Comment (omniaNS) [inline solutions/convert_temperature.py]: module docstrings are different from function docstrings, module docstrings describe the overall purpose of the file, while function docstrings describe individual functions. What I meant was to add a module docstring at the top of your files, like i described in my previous feedback.

**Review by omniaNS — APPROVED** [review]
